feat(core): Workflow Execution Statistics

freyamade opened the pull request.

Commit:
:tada: - Initial attempt at putting a DB table together

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
:sparkles: - Add statistics to DB.collections

Should allow for tests to run on PR

Commit:
:art: - actually run the statistics sqlite migration

Commit:
:sparkles: - add postgres migration

Commit:
:bug: - Fixed wrong table in down migrations

Commit:
:sparkles: - Added mysql migration

Commit:
:bug: - Postgres migration tested

Commit:
:sparkles: - Add dataLoaded flag to workflow table

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
:art: - Make latestEvent nullable

Commit:
:sparkles: - Load current execution counts during migration

Commit:
:bug: - Fixed issue with postgres migrations

Commit:
:art: - Better insertion queries for initial stats

krynble reviewed (changes requested):
![image](https://user-images.githubusercontent.com/219272/194869670-584ef53e-63d7-47b8-84f0-0a7151e7d721.png)

Hey @freyamade I've left a couple comments with suggestions.

Also I had the above issue when starting n8n. I can help you debug if you need, LMK.

Commit:
:art: - Remove initial counts of executions from migration

Commit:
:art: - Implemented change comments

Commit:
:sparkles: - Add new eventEmitter to core

Commit:
:sparkles: - Add workflow finished event handler

Commit:
:sparkles: - Trigger workflow finish events

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

csuermann commented on packages/cli/src/WorkflowExecuteAdditionalData.ts:
```diff
@@ -642,6 +642,8 @@ function hookFunctionsSave(parentProcessMode?: string): IWorkflowExecuteHooks {
 						);
 					}
+				} finally {
+					eventEmitter.emit('saveWorkflowStatistics', this.workflowData, fullRunData);
```
I'd suggest to introduce an event type enum instead of using magic strings. Also, events should be emitted when some relevant thing happened and denote that clearly in their name, e.g. "workflowExecutionCompleted" (as opposed to an instruction/command like "saveWorkflowStatistics" which would describe what (one of potentially many) event _listeners_ does.)

Commit:
:art: - Replaced magic strings with key/value pair

Commit:
:sparkles: - Add execution counts endpoint to internal api

Commit:
:sparkles: - Added timestamp endpoint

Commit:
:bug: - fix issue where new workflows had null for dataLoaded

Commit:
:sparkles: - Add dataLoaded endpoint

Commit:
:sparkles: - Add nodefetcheddata event handling

For some weird reason this doesn't work on node executions currently.
It did when it was first implemented and then it stopped.

Commit:
:sparkles: - set data loaded flag

Commit:
:sparkles: - workflow exec stats now have user id

ivov commented on packages/cli/src/telemetry/index.ts:
```diff
@@ -100,7 +105,9 @@ export class Telemetry {
 			const workflowId = properties.workflow_id;
 
-			this.executionCountsBuffer[workflowId] = this.executionCountsBuffer[workflowId] ?? {};
+			this.executionCountsBuffer[workflowId] = this.executionCountsBuffer[workflowId] ?? {
+				user_id: properties.user_id,
+			};
```
Here we are only sending `user_id` if `this.executionCountsBuffer[workflowId]` is nullish. Not familiar with the details of execution counts, but I expect in the normal (non-nullish) case the user ID won't be part of the track payload? Unless it's being already included.

Best way to confirm that it is working is in the PH dashboard.

freyamade replied:
Well my thought process was;
- `executionCountsBuffer` starts empty and is emptied after every pulse
- if the workflowId isn't present, which it won't be after being emptied, then it starts with a `user_id` being present when collecting its first execution count metric for the workflow
- f.ex. line 98 shows where the executionCountsBuffer is cleared again after a pulse is sent

I'll check the dashboard at least and see!

Commit:
:sparkles: - Add first prod success to posthog

Commit:
:sparkles: - Pass node to nodeFetchedData event

Commit:
:sparkles: - sending node fetch metrics to posthog

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
fix existing tests

Commit:
:sparkles: - Added tests for event handler code

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
:hammer: - Fixing config imports

flipswitchingmonkey commented on packages/cli/src/events/WorkflowStatistics.ts:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	// Try insertion and if it fails due to key conflicts then update the existing entry instead
+	try {
```
Could we use .upsert() here instead of waiting for an exception? (see https://typeorm.io/repository-api)

freyamade replied:
Upsert wasn't working when we tried it because the value that would be updated would be a count field that wouldn't initially exist for insert, since we have to start it at 1. It was something we did try before but I wasn't able to get it working

freyamade replied:
I think that was the issue with onConflict in queryBuilder, but upsert ended up being unusable since you can't use `set count = count + 1` for both insert and update which is why they have to be separate

krynble committed:
Merge branch 'master' into feature/workflow-execution-statistics

Commit:
:hammer: - fixed update issue for mysql

Commit:
:bug: - Fix first event having null timestamp

krynble reviewed (changes requested):
I was not able to test postgres since I don't have the DBMS set up locally right now, so I'd appreciate if you could also test this.

I've left a few comments with my findings.

freyamade commented:
Postgres worked before and works now as well for the issue you brought up yesterday evening @krynble

freyamade commented on packages/cli/src/databases/migrations/sqlite/1664196174000-WorkflowStatistics.ts:
```diff
@@ -0,0 +1,40 @@
+		await queryRunner.query(
+			`CREATE TABLE \`${tablePrefix}workflow_statistics\` (
+				"count" INTEGER DEFAULT 0,
+				"latestEvent" DATETIME,
```
I fixed it in SQLite with adding a @CreateDataColumn but that didn't fix MySQL or Postgres. It's only for the first entry that it's null, updates have it set as expected, so I may have to do some digging for this one

freyamade commented on packages/cli/src/databases/migrations/mysqldb/index.ts:
```diff
@@ -47,6 +48,7 @@ export const mysqlMigrations = [
 	AddNodeIds1658932910559,
 	AddJsonKeyPinData1659895550980,
 	CreateCredentialsUserRole1660062385367,
+	WorkflowStatistics1664196174002,
```
I moved it, can I leave the name the same or should I time-stamp?

Commit:
:bug: - Fix issue where creation timestamps were null

Commit:
:sparkles: - Add permission checking to stats apis

krynble committed:
Merge branch 'feature/workflow-execution-statistics' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
bring files up to date with recent refactors

Commit:
:arrow_up: - Updating posthog-node to latest

Commit:
:sparkles: - Ensuring backend pushes feat flags to PH

Commit:
:bug: - use getsharedworkflow function for checking permissions in api

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
:rotating_light: - fixing slight issue in tests

Commit:
:rotating_light: - more test fixes

krynble committed:
Merge branch 'feature/workflow-execution-statistics' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

krynble committed:
fix: Remove commented import

freyamade commented on packages/cli/src/api/workflowStats.api.ts:
```diff
@@ -0,0 +1,171 @@
+});
+
+// Helper function that validates the ID, throws an error if not valud
+async function checkWorkflowId(workflowId: string, user: User): Promise<WorkflowEntity> {
```
I was trying that but for some reason I was running into a weird issue, something along the lines of `column entityType does not exist` if I remember right :thinking:

Commit:
:hammer: - moved changeWorkflowID to middleware

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
Merge branch 'feature/workflow-execution-statistics' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into feature/workflow-execution-statistics

Commit:
:shirt: - clean up lint errors

Commit:
:shirt: - accidental typo

krynble commented on packages/cli/src/api/workflowStats.api.ts:
```diff
@@ -0,0 +1,171 @@
+	}
+
+	// Call the checkWorkflowId function here
+	await checkWorkflowId(req.params.id, req.user);
```
We can add this to another `workflowStatsController.use` call, so we have two separate middlewares: one for the logger and one for permissions.

One important note is that the current implementation does not prevent access, if it errors then it's simply ignored and I can access statistics about any workflow. We should instead use `res.send(...)` and prevent calling `next()` when there is any issue.

Commit:
:hammer: - Made checkWorkflowId work as middleware

Commit:
:bug: - next doesn't halt execution

krynble committed:
refactor: add brackets to improve legibility

Commit:
:shirt: - remove the name `err`

Commit:
Merge branch 'feature/workflow-execution-statistics' of github.com:n8n-io/n8n into feature/workflow-execution-statistics
